### PR TITLE
Suggestions

### PR DESCRIPTION
**Pull request by worldwideweary**
Here are some suggestions to the attempt to upstream to 4, but no guarantees.... but I suppose this might help 
Can't test the build unfortunately or anything of the sort, but the point is the form of the thing.

### CODE REVIEW

**Comment by worldwideweary** [comment]
Let me know if things make sense or if this even compiles. My system has all sorts of errors attempting to run cmake for MSS4, so I'm not currently set up for this stuff to help further

**Comment by Jojo-Schmitz** [comment]
Sure, just let me get my morning coffee first 😉

File: src/notation/internal/notationinteraction.cpp [hunk 1]
```diff
@@ -689,24 +689,70 @@ void NotationInteraction::moveChordNoteSelection(MoveDirection d)
...
+    std::vector<Element*> notes;
+
+    // Single traverse:
+    if (currentSingle && (currentSingle->isNote() || currentSingle->isRest())) {
+        EngravingItem* newSingle = score.moveAlt(currentSingle, d);
+        if (newSingle == currentSingle)
+            return;
+        while (newSingle && newSingle->isRest() && toRest(newSingle)->isGap()) {
+            newSingle = score.moveAlt(newSingle, d);
+            if (newSingle == old)
```
Comment (Jojo-Schmitz) [inline src/notation/internal/notationinteraction.cpp:712]: Shouldn#t this be
```cpp
            if (newSingle == currentSingle)
```

File: src/notation/internal/notationinteraction.cpp [hunk 2]
```diff
@@ -689,24 +689,70 @@ void NotationInteraction::moveChordNoteSelection(MoveDirection d)
...
+        auto it = std::unique(elements.begin(), elements.end());
+        elements.erase(it, elements.end());
+    };
 
-    EngravingItem* chordElem;
-    if (d == MoveDirection::Up) {
-        chordElem = score()->upAlt(current);
-    } else {
-        chordElem = score()->downAlt(current);
+    auto& score = *score();
```
Comment (Jojo-Schmitz) [inline src/notation/internal/notationinteraction.cpp:698]: Need to change `score` to something like `_score`

File: src/notation/internal/notationinteraction.cpp [hunk 3]
```diff
@@ -689,24 +689,70 @@ void NotationInteraction::moveChordNoteSelection(MoveDirection d)
...
+            }
+        select({ newSingle }, SelectType::SINGLE, newSingle->staffIdx());
+        showItem(newSingle);
+    }
+    // Range/List traverse:
+    else for (auto item : selection.elements()) {
+        if (item->isNote()) {
+            auto selectedNote = toNote(item);
+            if (isRange) {
+                auto newSelection = (d == Direction::DOWN)
```
Comment (Jojo-Schmitz) [inline src/notation/internal/notationinteraction.cpp:723]: Need to change `Direction::DOWN` to `MoveDirection::Down`

File: src/notation/internal/notationinteraction.cpp (hunk 1 above)
Comment (worldwideweary) [reply src/notation/internal/notationinteraction.cpp:712]: Oh, I think I meant oldSingle

Comment (worldwideweary) [reply src/notation/internal/notationinteraction.cpp:712]: it looks like it doesn't matter. Looks like currentSingle wasn't changed. So yeah currentSingle

File: src/notation/internal/notationinteraction.cpp [hunk 4]
```diff
@@ -689,24 +689,70 @@ void NotationInteraction::moveChordNoteSelection(MoveDirection d)
...
+                    ? selectedNote->chord()->downNote()
+                    : selectedNote->chord()->upNote();
+                notes.emplace_back(newSelection);
+            }
+            else { // List
+                auto newSelection = score.moveAlt(selectedNote, d);
```
Comment (Jojo-Schmitz) [inline src/notation/internal/notationinteraction.cpp:729]: More like `d == MoveDirection::Up ? DirectionV::UP : DirectionV::DOWN`

File: src/notation/internal/notationinteraction.cpp [hunk 5]
```diff
@@ -689,24 +689,70 @@ void NotationInteraction::moveChordNoteSelection(MoveDirection d)
...
+    bool isRange = selection.isRange();
+    std::vector<Element*> notes;
+
+    // Single traverse:
+    if (currentSingle && (currentSingle->isNote() || currentSingle->isRest())) {
+        EngravingItem* newSingle = score.moveAlt(currentSingle, d);
+        if (newSingle == currentSingle)
+            return;
+        while (newSingle && newSingle->isRest() && toRest(newSingle)->isGap()) {
+            newSingle = score.moveAlt(newSingle, d);
```
Comment (Jojo-Schmitz) [inline src/notation/internal/notationinteraction.cpp:711]: More like `d == MoveDirection::Up ? DirectionV::UP : DirectionV::DOWN`

File: src/notation/internal/notationinteraction.cpp [hunk 6]
```diff
@@ -689,24 +689,70 @@ void NotationInteraction::moveChordNoteSelection(MoveDirection d)
         return;
     }
 
-    EngravingItem* current = selection()->element();
-    if (!current || !(current->isNote() || current->isRest())) {
-        return;
-    }
+    auto removeDuplicates = [](std::vector<Element*>&elements) {
```
Comment (Jojo-Schmitz) [inline src/notation/internal/notationinteraction.cpp:692]: Changed `<Element*>` to `<EngravingItem*>`

File: src/notation/internal/notationinteraction.cpp [hunk 7]
```diff
@@ -689,24 +689,70 @@ void NotationInteraction::moveChordNoteSelection(MoveDirection d)
...
-    if (d == MoveDirection::Up) {
-        chordElem = score()->upAlt(current);
-    } else {
-        chordElem = score()->downAlt(current);
+    auto& score = *score();
+    auto& selection = score.selection();
+    EngravingItem* currentSingle = selection.element();
+    EngravingItem* oldSingle = currentSingle;
+    bool isRange = selection.isRange();
+    std::vector<Element*> notes;
```
Comment (Jojo-Schmitz) [inline src/notation/internal/notationinteraction.cpp:703]: Changed `<Element*>` to `<EngravingItem*>`

File: src/notation/internal/notationinteraction.cpp [hunk 8]
```diff
@@ -689,24 +689,70 @@ void NotationInteraction::moveChordNoteSelection(MoveDirection d)
...
+        }
+        removeDuplicates(notes);
     }
 
-    if (chordElem == current) {
+    if (!notes.empty()) {
+        selection.clear();
+        for (auto& note : notes) {
+            select({ note }, SelectType::ADD, note->staffIdx());
+            showItem(note);
```
Comment (Jojo-Schmitz) [inline src/notation/internal/notationinteraction.cpp:751]: This doesn't work, , need to change the type of `notes`, see above

**Comment by Jojo-Schmitz** [comment]
It doesn't compile, but should with just some minor changes

File: src/notation/internal/notationinteraction.cpp (hunk 4 above)
Comment (worldwideweary) [reply src/notation/internal/notationinteraction.cpp:729]: Looks like DirectionV type should be changed to MoveDirection for the function moveAlt() in the header

**Comment by worldwideweary** [comment]
What'd you do for the notes type

**Comment by worldwideweary** [comment]
I can try to push what you're doing here also real quick if it'd help

**Pull request merged**